File: homework/ba-nguyen/lesson-7/keys-and-rooms.ts
function canVisitAllRooms(rooms: number[][]): boolean {
  let count = 0;
  const visited: { [key: string]: boolean } = {};

  function dfs(roomIndex: number) {
    count++;
    visited[roomIndex] = true;

    for (const room of rooms[roomIndex]) {
      if (!visited[room]) {
        dfs(room);
      }
    }
  }

  dfs(0);

  return count === rooms.length;
}

File: homework/quang_tu/lesson_5/keys-and-rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        vector<bool> visited(rooms.size(), false);
        int count = 0;

        dfs(visited, rooms, 0, count);

        return count == rooms.size();
    }

    void dfs(vector<bool>& visited, vector<vector<int>>& rooms, int index, int& count) {
        visited[index] = true;
        count++;

        for (int i = 0; i < rooms[index].size(); ++i) {
            int val = rooms[index][i];

            if (!visited[val])
                dfs(visited, rooms, val, count);
        }
    }
};
